c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -1193,14 +1270,256 @@ class DeviceCachingAllocator {
     reset_peak_stat(stats.oversize_segments);
   }
 
+  /* Checkpoint the state of a private pool necessary to return it to its
+   * current state */
+  PrivatePoolState getCheckpointState(MempoolId_t id) {
+    std::lock_guard<std::recursive_mutex> lock(mutex);
+
+    auto pool = graph_pools.find(id);
+    if (pool != graph_pools.end()) {
+      auto pool_blocks = get_pool_blocks(pool->second.get());
+      return constructPrivatePoolState(pool->second.get(), id, pool_blocks);
+    } else if (graph_pools_freeable.count(id)) {
+      TORCH_CHECK(false, "Not expected to checkpoint freeable graph");
+    } else {
+      TORCH_CHECK(false, "Could not find pool of id");
+    }
+  }
+
+  void freeBlocksAllocatedToPool(PrivatePool* private_pool) {
+    std::unordered_map<void*, Block*> orig_ptrs_to_blocks;
+
+    auto pool_blocks = get_pool_blocks(private_pool);
+
+    std::vector<Block*> head_blocks;
+    for (Block* block : pool_blocks) {
+      if (block->prev == nullptr) {
+        head_blocks.push_back(block);
+      }
+    }
+
+    for (Block* block : head_blocks) {
+      Block* curr = block;
+
+      while (curr) {
+        // When we free a block, its pointer should never change
+        // only its adjacent blocks, so free, then look at pointer
+        if (curr->allocated) {
+          TORCH_CHECK(
```

Comment:
Is this a TORCH_CHECK or a TORCH_INTERNAL_ASSERT?

Reply:
i think it could be user visible 🤔 if you called get checkpoint in the middle of doing a cudagraph capture